File: app/src/main/java/opencrowd/hgc/hgcwallet/ui/main/account/AccountCreateScreen.java
package opencrowd.hgc.hgcwallet.ui.main.account;


import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.support.annotation.NonNull;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import com.wealthfront.magellan.BaseScreenView;
import com.wealthfront.magellan.Screen;

import opencrowd.hgc.hgcwallet.R;
import opencrowd.hgc.hgcwallet.common.Singleton;
import opencrowd.hgc.hgcwallet.database.DBHelper;
import opencrowd.hgc.hgcwallet.database.account.Account;
import opencrowd.hgc.hgcwallet.modals.HGCAccountID;
import opencrowd.hgc.hgcwallet.ui.auth.AuthActivity;
import opencrowd.hgc.hgcwallet.local_auth.AuthListener;
import opencrowd.hgc.hgcwallet.ui.customviews.TitleBarWrapper;

public class AccountCreateScreen extends Screen<AccountCreateView> implements AuthListener {

    Account account;
    private String titleName;
    private boolean image;

    public AccountCreateScreen(Account account,String title,boolean value) {
        this.account = account;
        this.titleName = title;
        this.image = value;
    }

    @NonNull
    @Override
    protected AccountCreateView createView(@NonNull Context context) {
        return new AccountCreateView(context,account,titleName,image);
    }

    @Override
    public void onAuthSetupSuccess() {

    }

    @Override
    public void onAuthSuccess(int requestCode) {
        if (requestCode == 100)
            getView().showPrivateKey();
    }

    @Override
    public void onAuthSetupFailed(boolean isCancelled) {

    }

    @Override
    public void onAuthFailed(int requestCode,boolean isCancelled) {

    }
}

class AccountCreateView extends BaseScreenView<AccountCreateScreen> {
    private EditText mEditTextNickName,mEditTextAccountId;
    private TextView mPublicKey,mPrivateKey,mDisplayText,mHideText;
    private Button mCancelbtn;
    private ImageView mCopyImageView,mCopyImageViewAccountData;
    @NonNull String publickey = "";
    private boolean isPrivateKeyDisplay = false;


    public AccountCreateView(@NonNull Context context, final @NonNull Account account, @NonNull String title, boolean imageValue) {
        super(context);
        inflate(context, R.layout.view_account_created_layout,this);
        TitleBarWrapper titleBar = new TitleBarWrapper(findViewById(R.id.titleBar));
        titleBar.setTitle(title);
       /* if(imageValue==true)
        {
            titleBar.setCloseButtonHidden(false); //if account created and account details icon
                                                      will be same then this code shoud be deleted
        }*/
       titleBar.setCloseButtonHidden(false);
        titleBar.setOnCloseButtonClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                getScreen().getNavigator().goBack();
            }
        });
        mEditTextNickName = (EditText)findViewById(R.id.edittext_nick_name);
        mEditTextAccountId = (EditText)findViewById(R.id.edittext_account_id);
        mPublicKey = (TextView)findViewById(R.id.textview_public_address);
        mPrivateKey = (TextView)findViewById(R.id.textview_private_key);
        mCancelbtn = (Button)findViewById(R.id.btn_close);
        mCopyImageView = (ImageView)findViewById(R.id.image_copy);
        mCopyImageViewAccountData = (ImageView)findViewById(R.id.image_copy_account_data);
        mDisplayText = (TextView)findViewById(R.id.text_display);
        mHideText = (TextView)findViewById(R.id.text_hide);
        if(account != null) {
            if(account.getName()!=null)
            {
                mEditTextNickName.setText(account.getName());
            }
            publickey = Singleton.INSTANCE.publicKeyString(account);
            if(publickey != null && !publickey.isEmpty()) {
                mPublicKey.setText(publickey);
            }

        }

        mCancelbtn.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                getScreen().getNavigator().goBack();
            }
        });

        if (account.accountID() != null) {
            mEditTextAccountId.setText(account.accountID().stringRepresentation());
        }

        mEditTextAccountId.addTextChangedListener(new TextWatcher() {

            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                HGCAccountID accountID = HGCAccountID.fromString(mEditTextAccountId.getText().toString());
                if (accountID != null) {
                    HGCAccountID existingAccID = account.accountID();
                    if (existingAccID != null) {
                        if (!existingAccID.equals(accountID)) {
                            Singleton.INSTANCE.clearAccountData(account);
                        }
                    }
                    account.setAccountID(accountID);
                    DBHelper.saveAccount(account);
                }
            }
        });

        mEditTextNickName.addTextChangedListener(new TextWatcher() {

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
              account.setName(mEditTextNickName.getText().toString());
              DBHelper.saveAccount(account);

            }
        });

        mCopyImageView.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                if(publickey!=null && !publickey.isEmpty()) {
                    Singleton.INSTANCE.copyToClipBoard(publickey,getScreen().getActivity());
                    Singleton.INSTANCE.showToast(getScreen().getActivity(), getScreen().getActivity().getString(R.string.copy_data_clipboard_message));

                }
            }
        });

        mCopyImageViewAccountData.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                String accountData = Singleton.INSTANCE.accountToJSONString(account, isPrivateKeyDisplay);
                Singleton.INSTANCE.copyToClipBoard(accountData,getScreen().getActivity());
                Singleton.INSTANCE.showToast(getScreen().getActivity(), getScreen().getActivity().getString(R.string.copy_data_clipboard_account_data));
            }
        });

        mDisplayText.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                new AlertDialog.Builder(getScreen().getActivity())
                        .setTitle("Warning")
                        .setMessage("There is some risk involved in showing the private key.\n\nDo you really want to see it?")
                        .setPositiveButton("No", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        dialogInterface.dismiss();
                    }
                }).setNegativeButton("Yes", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        isPrivateKeyDisplay = true ;
                        dialogInterface.dismiss();
                        AuthActivity authActivity = (AuthActivity)getScreen().getActivity();
                        authActivity.requestAuth(100);
                    }
                }).show();
            }
        });

        mHideText.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                isPrivateKeyDisplay = false ;
                mDisplayText.setVisibility(VISIBLE);
                mHideText.setVisibility(GONE);
                mPrivateKey.setText(R.string.default_private_key);
            }
        });
    }

    void showPrivateKey() {
        String privateKey = Singleton.INSTANCE.privateKeyString(getScreen().account);
        mDisplayText.setVisibility(GONE);
        mHideText.setVisibility(VISIBLE);
        mPrivateKey.setText(privateKey);
    }
}

File: app/src/main/java/hedera/hgc/hgcwallet/ui/main/account/AccountCreateScreen.kt
/*
 *
 *  Copyright 2019 Hedera Hashgraph LLC
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package hedera.hgc.hgcwallet.ui.main.account


import android.app.AlertDialog
import android.content.Context
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView

import com.wealthfront.magellan.BaseScreenView
import com.wealthfront.magellan.Screen

import hedera.hgc.hgcwallet.R
import hedera.hgc.hgcwallet.common.Singleton
import hedera.hgc.hgcwallet.database.DBHelper
import hedera.hgc.hgcwallet.database.account.Account
import hedera.hgc.hgcwallet.modals.HGCAccountID
import hedera.hgc.hgcwallet.ui.auth.AuthActivity
import hedera.hgc.hgcwallet.local_auth.AuthListener
import hedera.hgc.hgcwallet.ui.customviews.TitleBarWrapper

class AccountCreateScreen(account: Account, titleName: String, image: Boolean) : Screen<AccountCreateView>(), AuthListener {

    data class Params(val account: Account, val title: String, val imageValue: Boolean, var shouldShowPrivateKey: Boolean)

    private val params = Params(account, titleName, image, false)
    override fun createView(context: Context): AccountCreateView {
        return AccountCreateView(context, params)
    }

    override fun onAuthSetupSuccess() {
    }

    override fun onAuthSuccess(requestCode: Int) {
        if (requestCode == 100)
            view?.reload()

    }

    override fun onAuthSetupFailed(isCancelled: Boolean) {
    }

    override fun onAuthFailed(requestCode: Int, isCancelled: Boolean) {
    }

    internal fun displayPrivateKey(show: Boolean) {
        params.shouldShowPrivateKey = show
        if (show)
            (activity as? AuthActivity)?.requestAuth(100)
    }

    internal fun onCopyPublicKeyClick() {
        val publickey = Singleton.publicKeyString(params.account)
        val privatekey = Singleton.privateKeyString(params.account)
        Singleton.copyToClipBoard("$publickey || $privatekey", activity)
        Singleton.showToast(activity, activity.getString(R.string.copy_data_clipboard_message))
    }

    internal fun onCopyAccountDataClick() {
        Singleton.accountToJSONString(params.account, params.shouldShowPrivateKey)?.let { accountData ->
            Singleton.copyToClipBoard(accountData, activity)
            Singleton.showToast(activity, activity.getString(R.string.copy_data_clipboard_account_data))
        }
    }

    internal fun updateAccountName(text: String) {
        params.account.name = text
        DBHelper.saveAccount(params.account)
    }

    internal fun updateAccountId(text: String) {
        HGCAccountID.fromString(text)?.let { accountID ->
            params.account.accountID()?.let {
                if (it != accountID)
                    Singleton.clearAccountData(params.account)
            }
            params.account.setAccountID(accountID)
            DBHelper.saveAccount(params.account)
        }
    }
}

class AccountCreateView(context: Context, private val params: AccountCreateScreen.Params) : BaseScreenView<AccountCreateScreen>(context) {
    private val mEditTextNickName: EditText
    private val mEditTextAccountId: EditText
    private val mPublicKey: TextView
    private val mPrivateKey: TextView
    private val mDisplayText: TextView
    private val mHideText: TextView
    private val mCopyImageView: ImageView
    private val mCopyImageViewAccountData: ImageView

    init {
        View.inflate(context, R.layout.view_account_created_layout, this)
        TitleBarWrapper(findViewById(R.id.titleBar)).apply {
            setTitle(params.title)
            setCloseButtonHidden(false)
            setOnCloseButtonClickListener { screen.navigator.goBack() }
        }

        mEditTextNickName = findViewById<EditText>(R.id.edittext_nick_name)
        mEditTextAccountId = findViewById<EditText>(R.id.edittext_account_id)
        mPublicKey = findViewById<TextView>(R.id.textview_public_address)
        mPrivateKey = findViewById(R.id.textview_private_key)
        mCopyImageView = findViewById<ImageView>(R.id.image_copy)
        mCopyImageViewAccountData = findViewById<ImageView>(R.id.image_copy_account_data)
        mDisplayText = findViewById<TextView>(R.id.text_display)
        mHideText = findViewById<TextView>(R.id.text_hide)


        findViewById<Button>(R.id.btn_close).setOnClickListener { screen.navigator.goBack() }


        mEditTextAccountId.addTextChangedListener(object : TextWatcher {

            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
            }

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
            }

            override fun afterTextChanged(editable: Editable) {
                screen?.updateAccountId(mEditTextAccountId.text.toString())
            }
        })

        mEditTextNickName.addTextChangedListener(object : TextWatcher {

            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            }

            override fun afterTextChanged(s: Editable) {
                screen?.updateAccountName(mEditTextNickName.text.toString())
            }
        })

        mCopyImageView.setOnClickListener {
            val publicKey = mPublicKey.text.toString()
            if (publicKey.isNotEmpty())
                screen?.onCopyPublicKeyClick()

        }

        mCopyImageViewAccountData.setOnClickListener {
            screen?.onCopyAccountDataClick()
        }

        mDisplayText.setOnClickListener {
            AlertDialog.Builder(screen.activity)
                    .setTitle("Warning")
                    .setMessage("There is some risk involved in showing the private key.\n\nDo you really want to see it?")
                    .setPositiveButton("No") { dialogInterface, i -> dialogInterface.dismiss() }.setNegativeButton("Yes") { dialogInterface, i ->
                        dialogInterface.dismiss()
                        screen?.displayPrivateKey(true)
                    }.show()
        }

        mHideText.setOnClickListener {
            screen?.displayPrivateKey(false)
            reload()
        }

        reload()
    }

    fun reload() {

        params.account.let {
            mEditTextNickName.setText(it.name)
            val publickey = Singleton.publicKeyString(it)
            if (publickey.isNotEmpty())
                mPublicKey.text = publickey
        }

        params.account.accountID()?.let {
            mEditTextAccountId.setText(it.stringRepresentation())
        }

        if (params.shouldShowPrivateKey) {
            val privateKey = Singleton.privateKeyString(params.account)
            mDisplayText.visibility = View.GONE
            mHideText.visibility = View.VISIBLE
            mPrivateKey.text = privateKey
        } else {
            mDisplayText.visibility = View.VISIBLE
            mHideText.visibility = View.GONE
            mPrivateKey.setText(R.string.default_private_key)
        }
    }
}
